src/Interface/ShapeHelperOpInterface.hpp

```diff
@@ -0,0 +1,143 @@
+
+   @param op Operation to be analyzed.
+
+   @param operands Operands of the operation to be analyzed. When passing an
+   empty list, the operands are taken from the operations. When passing an
+   explicit, non-empty list, these operands are used instead of the ones from
+   the operation.
+
+   The former option (empty list) is used during shape inference.
+
+   The later option (explicit list) is used during lowering, as typically there
+   are two sets of operands, the ones already lowered (explicit list) and the
+   original ones (contained within the op). Using shapes during lowering
+   typically deals with already-lowered operands.
+
+   However, during lowering, it may be sometime advantageous to perform the
+   analysis of the index expressions in the "old" dialect, e.g. in ONNX instead
+   of the destination dialect. To enable this, just pass `{}` as operands and
+   the original operands associated with the unmodified operation will be used.
+
+   @param ieBuilder Class that scans the operands to gather IndexExpr from them.
+   Typically used to gather shape and constant values.
+
+   During shape inference, we typically use IndexExprBuilderForAnalysis
+   (src/Dialect/Mlir/DialectBuilder.hpp), which uses questionmark for values
+   unkown at compile time. This builder is default when no ieBuilder is given.
+
+   During lowering, we typically use and Index Expr Builder that generates code
+   for values unknown at compile time. Example of such subclasses are
+   IndexExprBuilderForKrnl (generates Krnl ops, in
+   src/Dialect/Krnl/DialectBuilder.hpp, ) or IndexExprBuilderForMhlo (generates
+   Shape/MHLO ops, in src/Conversion/ONNXToMhlo/DialectBuilder.hpp).
+
+   @param scope Index expression scope to be used. If none is provided, a new
+   scope is created and stored internally. This scope will then be destructed
+   when the current object is destructed.
+
+   Passing a scope is critically important when, to evaluate the shape of a
+   given operation, we must also analysis the shape of an other operation. Both
+   shape helper MUST share the same scope as otherwise there will be references
+   to "deleted" index expressions (as all index expressions are deleted when its
+   directly enclosing scope vanishes).
+   */
+
+  ONNXOpShapeHelper(mlir::Operation *op,    /* Op to be analyzed. */
+      mlir::ArrayRef<mlir::Value> operands, /* If empty, use operands from op.*/
+      IndexExprBuilder *ieBuilder, /* Use IndexExprBuilderForAnalysis if null.*/
+      IndexExprScope *scope);      /* Install local scope if null. */
+  virtual ~ONNXOpShapeHelper();
+
+  // Every leaf class is expected to create a computeShape with the following
+  // signature. This method is responsible to compute at a minimum the output
+  // dims.
+  virtual mlir::LogicalResult computeShape() = 0;
+  // Helper function that set n'th output dims from the given value.
+  mlir::LogicalResult computeShapeFromOperand(mlir::Value operand, int n = 0);
```

Comment:
The purpose of this function is unclear to me: Can the shape be computed with only one operand? Should "ValueRange operands" be used or does this function is just to set the dim?

Reply:
@chentong319 some ONNX ops have multiple inputs but the shape is only defined by one operand (meaning identical to that one). This function allow me to reuse the "set output # `i` to be the same as the shape as the type of value `operand`". This only work for some ops, obviously.

Reply:
Based on your input, I am clarifying the comments in a subsequent PR, thanks for the feedback on this one.